packages/xsnap/src/xsnap.js

```diff
@@ -0,0 +1,234 @@
+// @ts-check
+/* eslint no-await-in-loop: ["off"] */
+
+/**
+ * @typedef {typeof import('child_process').spawn} Spawn
+ */
+
+/**
+ * @template T
+ * @typedef {import('./defer').Deferred<T>} Deferred
+ */
+
+import { defer } from './defer';
+import * as netstring from './netstring';
+import * as node from './node-stream';
+
+const OK = '.'.charCodeAt(0);
+const ERROR = '!'.charCodeAt(0);
+const QUERY = '?'.charCodeAt(0);
+
+const importMetaUrl = `file://${__filename}`;
+
+const encoder = new TextEncoder();
+const decoder = new TextDecoder();
+
+/**
+ * @param {Uint8Array} arg
+ * @returns {Uint8Array}
+ */
+function echoCommand(arg) {
+  return arg;
+}
+
+/**
+ * @param {Object} options
+ * @param {string} options.os
+ * @param {Spawn} options.spawn
+ * @param {(request:Uint8Array) => Promise<Uint8Array>} [options.handleCommand]
+ * @param {string=} [options.name]
+ * @param {boolean=} [options.debug]
+ * @param {string=} [options.snapshot]
+ * @param {'ignore' | 'inherit'} [options.stdout]
+ * @param {'ignore' | 'inherit'} [options.stderr]
+ */
+export function xsnap(options) {
+  const {
+    os,
+    spawn,
+    name = '<unnamed xsnap worker>',
+    handleCommand = echoCommand,
+    debug = false,
+    snapshot = undefined,
+    stdout = 'inherit',
+    stderr = 'inherit',
+  } = options;
+
+  const platform = {
+    Linux: 'lin',
+    Darwin: 'mac',
+    Windows_NT: 'win',
+  }[os];
+
+  if (platform === undefined) {
+    throw new Error(`xsnap does not support platform ${os}`);
+  }
+
+  const xsnapBin = new URL(
+    `../build/bin/${platform}/release/xsnap`,
+    importMetaUrl,
+  ).pathname;
+  const xsnapDebugBin = new URL(
+    `../build/bin/${platform}/debug/xsnap`,
+    importMetaUrl,
+  ).pathname;
+
+  /** @type {Deferred<Error?>} */
+  const vatExit = defer();
+
+  const args = snapshot ? ['-r', snapshot] : [];
+
+  const bin = debug ? xsnapDebugBin : xsnapBin;
+
+  const xsnapProcess = spawn(bin, args, {
```

Comment:
do we get an actionable error message if the binary wasn't compiled first? or is that not even a thing, since it will always get compiled in the process of installing this file? I'm not sure how install-time scripts will work in the monorepo symlinking step.

Reply:
I think so.

I got ENOENT aka file not found once before I fixed a typo in the lin makefiles.

Reply:
Yes, it fails but not mysteriously.